events/reporter.go

```diff
@@ -102,18 +110,13 @@ func ReportRewardReceived(r Reward) {
 		SmesherID: r.Smesher.ToBytes(),
 	})
 
-	if reporter != nil {
-		if reporter.blocking {
-			reporter.channelReward <- r
-			log.Debug("reported reward: %v", r)
-		} else {
-			select {
-			case reporter.channelReward <- r:
-				log.Debug("reported reward: %v", r)
-			default:
-				log.Debug("not reporting reward as no one is listening: %v", r)
-			}
+	if reporter != nil && len(reporter.rewardsSubs) > 0 {
+		log.Info("about to report reward: %v", r)
+		for _, sub := range reporter.rewardsSubs {
+			sub <- r
```

Comment:
Need to use a `select{}` statement here to make the send non-blocking - it must never block. And print a Debug/Info if no subscribers, or if any of the channels is full.

Reply:
I think that this means that if the channel is full because the client is slow pulling rewards from it (super slow network connection), then the client will miss reported rewards. Ideally, slow clients won't miss any events.

Reply:
Also - won't this code block in line 121 in case that the channel is full?

Reply:
Didn't you mean to remove this line as the send happens in line 117?

Reply:
Yes, I meant to remove the line apologies. If we don't want slow clients to miss rewards, should we then try sends in goroutines instead? So that the reporter itself won't block, but the individual goroutine will, until the client is ready

Reply:
I think we should but it needs to be done with care because the subscribers list is reporter state and can change. So, I think that each specific send needs to be done in a go routine and not 1 go routine with iteration over subscribers list inside of it. Also curious what @noamnelke thinks on the best design for this.

Reply:
The solution here is _not_ one goroutine per send. This is exactly the use case for buffered channels. Just set a capacity for the channel, write to the channel, and let the consumer consume it. If the channel is full, the write should not block, it should drop the data (and print a log).

Reply:
Also - we want to apply this not just to rewards but to all streaming apis that smapp 0.2 needs.

Reply:
yes, please go over api services definitions in the api repo and collect all streaming methods. For example, NodeService has 2 important ones. All should support multiple slow clients.